community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java

```diff
@@ -1051,22 +1050,54 @@ void readIntoRecord( long id, RECORD record, RecordLoad mode, PageCursor cursor
         int offset = offsetForId( id );
         if ( cursor.next( pageId ) )
         {
-            // There is a page in the store that covers this record, go read it
-            do
-            {
-                prepareForReading( cursor, offset, record );
-                recordFormat.read( record, cursor, mode, recordSize );
-            }
-            while ( cursor.shouldRetry() );
-            checkForDecodingErrors( cursor, id, mode );
-            verifyAfterReading( record, mode );
+            readRecordFromPage( id, record, mode, cursor, offset );
         }
         else
         {
             verifyAfterNotRead( record, mode );
         }
     }
 
+    @Override
+    public void nextRecordByCursor( RECORD record, RecordLoad mode, PageCursor cursor ) throws UnderlyingStorageException
+    {
+        try
+        {
+            int offset = cursor.getOffset() + recordSize;
+            long id = record.getId() + 1;
+            record.setId( id );
+            long pageId = cursor.getCurrentPageId();
+            if ( offset >= storeFile.pageSize() || pageId < 0 )
+            {
+                offset = 0;
+                if ( !cursor.next( pageId + 1 ) )
```

Comment:
`pageId + 1` should really be `0`, right?

Reply:
That depends on which if condition that got us into this block. Now it is a little bit sneakily depending on `pageId` being -1 when it's not initialized, which I guess should be made more explicit.

Reply:
Oh, right. Clever, but please add a comment :)